tests/test_pca.py

```diff
@@ -83,75 +81,105 @@ def wrapper(a: np.ndarray) -> DaskArray:
-    ]
-)
-def array_type(request: pytest.FixtureRequest):
+def conv_at(array_type):
     # If one uses dask for PCA it will always require dask-ml.
     # dask-ml can’t do 2D-chunked arrays, so rechunk them.
-    if as_dask_array := DASK_CONVERTERS.get(request.param):
-        return as_dask_array
+    if as_dask_array := DASK_CONVERTERS.get(array_type):
+        return (as_dask_array,)
 
     # When not using dask, just return the array type
-    assert "dask" not in request.param.__name__, "add more branches or refactor"
-    return request.param
+    assert "dask" not in array_type.__name__, "add more branches or refactor"
+    return (array_type,)
 
 
-@pytest.fixture(params=[None, "valid", "invalid"])
-def svd_solver_type(request: pytest.FixtureRequest):
-    return request.param
+ARRAY_TYPES = [
+    param_with(at, conv_at, marks=[needs.dask_ml]) if "dask" in cast(str, at.id) else at
+    for at in ARRAY_TYPES_SPARSE_DASK_UNSUPPORTED
+]
 
 
-@pytest.fixture(params=[True, False], ids=["zero_center", "no_zero_center"])
-def zero_center(request: pytest.FixtureRequest):
+@pytest.fixture(params=ARRAY_TYPES)
+def array_type(request: pytest.FixtureRequest):
     return request.param
 
 
-@pytest.fixture
-def pca_params(
-    array_type, svd_solver_type: Literal[None, "valid", "invalid"], zero_center
-):
-    all_svd_solvers = {"auto", "full", "arpack", "randomized", "tsqr", "lobpcg"}
-
-    warn_pat_expected = None
-    svd_solver = None
-    if svd_solver_type is not None:
-        match array_type, zero_center:
-            case (dc, True) if dc in DASK_CONVERTERS.values():
-                svd_solver = {"auto", "full", "tsqr", "randomized"}
-            case (dc, False) if dc in DASK_CONVERTERS.values():
-                svd_solver = {"tsqr", "randomized"}
-            case ((sparse.csr_matrix | sparse.csc_matrix), True):
-                svd_solver = {"arpack"}
-            case ((sparse.csr_matrix | sparse.csc_matrix), False):
-                svd_solver = {"arpack", "randomized"}
-            case (helpers.asarray, True):
-                svd_solver = {"auto", "full", "arpack", "randomized"}
-            case (helpers.asarray, False):
-                svd_solver = {"arpack", "randomized"}
-            case _:
-                pytest.fail(f"Unknown array type {array_type}")
-        if svd_solver_type == "invalid":
-            svd_solver = all_svd_solvers - svd_solver
-            warn_pat_expected = r"Ignoring"
-
-        svd_solver = random.choice(list(svd_solver))
-    # explicit check for special case
-    if (
-        array_type in {sparse.csr_matrix, sparse.csc_matrix}
-        and zero_center
-        and svd_solver == "lobpcg"
-    ):
-        warn_pat_expected = r"legacy code"
+SVDSolver = Literal["auto", "full", "arpack", "randomized", "tsqr", "lobpcg"]
+
 
-    return (svd_solver, warn_pat_expected)
+def gen_pca_params(
+    *, array_type, svd_solver_type: Literal[None, "valid", "invalid"], zero_center: bool
```

Comment:
Ok, so there's an issue with typing `array_type`?

Reply:
no, I can do it!